components/sup/src/event/nats_message_stream.rs

```diff
@@ -1,30 +1,116 @@
-use crate::event::{Error,
-                   EventStreamConfig,
+use crate::event::{EventStreamConfig,
                    Result};
 use futures::{channel::{mpsc as futures_mpsc,
                         mpsc::UnboundedSender},
               stream::StreamExt};
-use rants::{error::Error as RantsError,
-            native_tls::TlsConnector,
-            Client,
-            Subject};
-use tokio::time;
+use nats::{self,
+           asynk::{Options, Connection}};
+use tokio::{time, sync::Mutex};
+use std::{
+    io,
+    path::PathBuf,
+    sync::Arc,
+    time::Duration};
 
 /// The subject and payload of a NATS message.
 #[derive(Debug)]
 pub struct NatsMessage {
-    subject: &'static Subject,
+    subject: &'static str,
     payload: Vec<u8>,
 }
 
 impl NatsMessage {
-    pub fn new(subject: &'static Subject, payload: Vec<u8>) -> Self {
+    pub fn new(subject: &'static str, payload: Vec<u8>) -> Self {
         NatsMessage { subject, payload }
     }
 
     pub fn payload(&self) -> &[u8] { self.payload.as_slice() }
 }
 
+#[derive(Clone)]
+pub struct NatsClient(Option<Connection>);
+
+impl NatsClient {
+    pub async fn connect(this: Arc<Mutex<Self>>, supervisor_id: String, config: EventStreamConfig) {        
+        match config.connect_method.into() {
+            Some(timeout) => {
+                trace!("Timeout used -> {:?}", timeout);
+                if let Err(e) = time::timeout(timeout, this.lock().await.connect_impl_with_reconnect(supervisor_id, &config))
+                    .await {
+                    trace!("Error connecting: {}", e);
+                    //  The process does not exit.  We should shut down if timeout specified and we cannot connect.
+                    //std::process::exit(1);
+                }
+            } 
+            None => {
+                trace!("Timeout not used");
+                tokio::spawn(async move { this.lock().await.connect_impl(supervisor_id, &config).await });
+            }
+        };
+    }
+
+    async fn connect_impl_with_reconnect(&mut self, supervisor_id: String, config: &EventStreamConfig) {        
+        while self.0.is_none() {
+            match Self::options_from_config(&supervisor_id, config)
+                .with_retry_first_connect()
```

Comment:
I would move these options builder calls to `options_from_config`.

Reply:
Updated